mathics/builtin/drawing/discrete_plot.py

```diff
@@ -0,0 +1,179 @@
+    }
+
+    def eval(self, function, x, start, nmax, step, evaluation, options):
+        """DiscretePlot[function_, {x_Symbol, start_Integer, nmax_Integer, step_Integer},
+        OptionsPattern[DiscretePlot]]"""
+        if isinstance(function, Symbol) and function.name is not x.get_name():
+            rules = evaluation.definitions.get_ownvalues(function.name)
+            for rule in rules:
+                functions = rule.apply(function, evaluation, fully=True)
+
+        if function.get_head_name() == "List":
+            functions_param = self.get_functions_param(functions)
+            for index, f in enumerate(functions_param):
+                if isinstance(f, Symbol) and f.name is not x.get_name():
+                    rules = evaluation.definitions.get_ownvalues(f.name)
+                    for rule in rules:
+                        f = rule.apply(f, evaluation, fully=True)
+                functions_param[index] = f
+            functions = functions.flatten_with_respect_to_head(SymbolList)
+
+        expr_limits = ListExpression(x, start, nmax)
+        # FIXME: arrange for self to have a .symbolname property or attribute
+        expr = Expression(
+            Symbol(self.get_name()), function, expr_limits, *options_to_rules(options)
+        )
+        function = self.get_functions_param(function)
+        x_name = x.get_name()
+
+        py_start = start.value
+        py_nmax = nmax.value
+        py_step = step.value
+        if py_start is None or py_nmax is None:
+            return evaluation.message(self.get_name(), "plln", nmax, expr)
+        if py_start >= py_nmax:
+            return evaluation.message(self.get_name(), "plld", expr_limits)
+
+        plotrange_option = self.get_option(options, "PlotRange", evaluation)
+        plot_range = eval_N(plotrange_option, evaluation).to_python()
+
+        x_range, y_range = self.get_plotrange(plot_range, py_start, py_nmax)
+        if not check_plot_range(x_range, numbers.Real) or not check_plot_range(
+            y_range, numbers.Real
+        ):
+            evaluation.message(self.get_name(), "prng", plotrange_option)
+            x_range, y_range = [py_start, py_nmax], "Automatic"
+
+        # x_range and y_range are now either Automatic, All, or of the form [min, max]
+        assert x_range in ("System`Automatic", "System`All") or isinstance(
+            x_range, list
+        )
+        assert y_range in ("System`Automatic", "System`All") or isinstance(
+            y_range, list
+        )
+
+        # PlotPoints Option
+        # constants to generate colors
+
+        base_plot_points = []  # list of points in base subdivision
+        plot_points = []  # list of all plotted points
+
+        compiled_fn = compile_quiet_function(
+            function[0], [x_name], evaluation, self.expect_list
+        )
+
+        @lru_cache
+        def apply_fn(fn, x_value: int) -> Optional[float]:
```

Comment:
One of the things I find amusing is that in the old code this is called "eval_fn". So not only did it sometimes use "apply" when "eval" was meant, but it also used "eval" when "apply" was meant.